File: android/electrode-reactnative-bridge/src/androidTest/java/com/walmartlabs/electrode/reactnative/sample/api/PersonRequests.java
/*
 * Copyright 2017 WalmartLabs

 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at

 * http://www.apache.org/licenses/LICENSE-2.0

 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.walmartlabs.electrode.reactnative.sample.api;

import android.support.annotation.NonNull;

import com.walmartlabs.electrode.reactnative.bridge.ElectrodeBridgeRequestHandler;
import com.walmartlabs.electrode.reactnative.bridge.ElectrodeBridgeResponseListener;
import com.walmartlabs.electrode.reactnative.bridge.None;
import com.walmartlabs.electrode.reactnative.bridge.RequestHandlerHandle;
import com.walmartlabs.electrode.reactnative.bridge.RequestHandlerProcessor;
import com.walmartlabs.electrode.reactnative.bridge.RequestProcessor;
import com.walmartlabs.electrode.reactnative.sample.model.Person;
import com.walmartlabs.electrode.reactnative.sample.model.Status;

import java.util.List;

/***
 * This is a test API generated for testing the bridge. This is first generated using api gen commend and necessary changes are made.
 * <p>
 * This needs to be kept as a reference for generating the api code. This is how a generated API code looks like(format, order, naming conventions etc.)
 * <p>
 * This class provides all the request actions that can be performed on Person.
 */
final class PersonRequests implements PersonApi.Requests {

    PersonRequests() {

    }

    @Override
    public RequestHandlerHandle registerGetPersonRequestHandler(@NonNull final ElectrodeBridgeRequestHandler<None, Person> handler) {
        return new RequestHandlerProcessor<>(REQUEST_GET_PERSON, None.class, Person.class, handler).execute();
    }

    @Override
    public RequestHandlerHandle registerGetStatusRequestHandler(@NonNull final ElectrodeBridgeRequestHandler<Person, Status> handler) {
        return new RequestHandlerProcessor<>(REQUEST_GET_STATUS, Person.class, Status.class, handler).execute();
    }

    @Override
    public RequestHandlerHandle registerGetAgeRequestHandler(@NonNull ElectrodeBridgeRequestHandler<String, Integer> handler) {
        return new RequestHandlerProcessor<>(REQUEST_GET_AGE, String.class, Integer.class, handler).execute();
    }

    @Override
    public RequestHandlerHandle registerUpdatePersonRequestHandler(@NonNull ElectrodeBridgeRequestHandler<UpdatePersonRequestData, Person> handler) {
        return new RequestHandlerProcessor<>(REQUEST_POST_PERSON_UPDATE, UpdatePersonRequestData.class, Person.class, handler).execute();
    }

    @Override
    public RequestHandlerHandle registerFindPersonsByStatus(@NonNull ElectrodeBridgeRequestHandler<List<Status>, List<Person>> handler) {
        return new RequestHandlerProcessor<>(REQUEST_FIND_PERSONS_BY_STATUS, (Class) Status.class, (Class) Person.class, handler).execute();
    }

    @Override
    public RequestHandlerHandle registerFindPersonsAgeByName(@NonNull ElectrodeBridgeRequestHandler<List<String>, List<Integer>> handler) {
        return new RequestHandlerProcessor<>(REQUEST_FIND_PERSONS_AGE_BY_NAME, (Class) String.class, (Class) Integer.class, handler).execute();
    }

    @Override
    public void getPerson(@NonNull final ElectrodeBridgeResponseListener<Person> responseListener) {
        new RequestProcessor<>(REQUEST_GET_PERSON, null, Person.class, responseListener).execute();
    }


    @Override
    public void getStatus(@NonNull Person person, @NonNull final ElectrodeBridgeResponseListener<Status> responseListener) {
        new RequestProcessor<>(REQUEST_GET_STATUS, person, Status.class, responseListener).execute();
    }


    @Override
    public void getUserName(@NonNull final ElectrodeBridgeResponseListener<String> responseListener) {
        new RequestProcessor<None, String>(REQUEST_GET_USER_NAME, null, String.class, responseListener).execute();
    }

    @Override
    public void getAge(@NonNull String name, @NonNull final ElectrodeBridgeResponseListener<Integer> responseListener) {
        new RequestProcessor<>(REQUEST_GET_AGE, name, Integer.class, responseListener).execute();
    }

    @Override
    public void updatePersonPost(@NonNull UpdatePersonRequestData updatePersonRequestData, @NonNull final ElectrodeBridgeResponseListener<Person> responseListener) {
        new RequestProcessor<>(REQUEST_POST_PERSON_UPDATE, updatePersonRequestData, Person.class, responseListener).execute();
    }

    @Override
    public void findPersonsByStatus(@NonNull List<Status> statusList, @NonNull ElectrodeBridgeResponseListener<List<Person>> responseListener) {
        new RequestProcessor<>(REQUEST_FIND_PERSONS_BY_STATUS, statusList, (Class) List.class, Person.class, responseListener).execute();
    }

    @Override
    public void findPersonsAgeByName(@NonNull List<String> names, @NonNull ElectrodeBridgeResponseListener<List<Integer>> responseListener) {
        new RequestProcessor<>(REQUEST_FIND_PERSONS_AGE_BY_NAME, names, (Class) List.class, Integer.class, responseListener).execute();
    }
}

File: ios/ElectrodeReactNativeBridgeTests/PersonRequests.swift
/*
 * Copyright 2017 WalmartLabs
 
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 
 * http://www.apache.org/licenses/LICENSE-2.0
 
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import UIKit

class PersonRequests: PersonAPI.Requests {
    override func registerGetPersonRequestHandler(handler: @escaping ElectrodeBridgeRequestCompletionHandler) {
        let requestHandlerProcessor = ElectrodeRequestHandlerProcessor(requestName: PersonAPI.kRequestGetPerson, reqClass: None.self, respClass: Person.self, requestCompletionHandler: handler)
        requestHandlerProcessor.execute()
    }

    override func registerGetStatusRequestHandler(handler: @escaping ElectrodeBridgeRequestCompletionHandler) {
        let requestHandlerProcessor = ElectrodeRequestHandlerProcessor(requestName: PersonAPI.kRequestGetStatus, reqClass: Person.self, respClass: Status.self, requestCompletionHandler: handler)
        requestHandlerProcessor.execute()
    }

    override func registerGetAgeRequestHandler(handler: @escaping ElectrodeBridgeRequestCompletionHandler) {
        let requestHandler = ElectrodeRequestHandlerProcessor(requestName: PersonAPI.kRequestGetAge, reqClass: String.self, respClass: Int.self, requestCompletionHandler: handler)
        requestHandler.execute()
    }

    override func registerUpdatePersonRequestHandler(handler: @escaping ElectrodeBridgeRequestCompletionHandler) {
        let requestHandler = ElectrodeRequestHandlerProcessor(requestName: PersonAPI.kRequestPostPersonUpdate, reqClass: UpdatePersonRequestData.self, respClass: Person.self, requestCompletionHandler: handler)
        requestHandler.execute()
    }

    override func registerFindPersonsByStatus(handler: @escaping ElectrodeBridgeRequestCompletionHandler) {
        let requestHandler = ElectrodeRequestHandlerProcessor(requestName: PersonAPI.kRequestFindPersonsByStatus, reqClass: Array<Any>.self, reqItemType: Status.self, respClass: Person.self, requestCompletionHandler: handler)
        requestHandler.execute()
    }

    override func registerFindPersonsAgeByName(handler: @escaping ElectrodeBridgeRequestCompletionHandler) {
        let requestHandler = ElectrodeRequestHandlerProcessor(requestName: PersonAPI.kRequestFindPersonsByName, reqClass: String.self, respClass: Int.self, requestCompletionHandler: handler)
        requestHandler.execute()
    }

    override func getPerson(responseCompletionHandler: @escaping ElectrodeBridgeResponseCompletionHandler) {
        let requestProcessor = ElectrodeRequestProcessor<Bridgeable, Person, Any>(requestName: PersonAPI.kRequestGetPerson,
                                                                                  requestPayload: nil,
                                                                                  respClass: Person.self,
                                                                                  responseItemType: nil,
                                                                                  responseCompletionHandler: responseCompletionHandler)
        requestProcessor.execute()
    }

    override func getStatus(person: Person,
                            responseCompletionHandler: @escaping ElectrodeBridgeResponseCompletionHandler) {
        let requestProcessor = ElectrodeRequestProcessor<Person, Status, Any>(requestName: PersonAPI.kRequestGetStatus,
                                                                              requestPayload: person,
                                                                              respClass: Status.self,
                                                                              responseItemType: nil,
                                                                              responseCompletionHandler: responseCompletionHandler)
        requestProcessor.execute()
    }

    override func getUserName(responseCompletionHandler: @escaping ElectrodeBridgeResponseCompletionHandler) {
        let requestProcessor = ElectrodeRequestProcessor<None, String, Any>(requestName: PersonAPI.kRequestGetUserName,
                                                                            requestPayload: nil,
                                                                            respClass: String.self,
                                                                            responseItemType: nil,
                                                                            responseCompletionHandler: responseCompletionHandler)
        requestProcessor.execute()
    }

    override func getAge(name: String, responseCompletionHandler: @escaping ElectrodeBridgeResponseCompletionHandler) {
        let requestProcessor = ElectrodeRequestProcessor<None, String, Any>(requestName: PersonAPI.kRequestGetAge,
                                                                            requestPayload: name,
                                                                            respClass: String.self,
                                                                            responseItemType: nil,
                                                                            responseCompletionHandler: responseCompletionHandler)
        requestProcessor.execute()
    }

    override func updatePersonPost(updatePersonRequestData: UpdatePersonRequestData,
                                   responseCompletionHandler: @escaping ElectrodeBridgeResponseCompletionHandler) {
        let requestProcessor = ElectrodeRequestProcessor<Bridgeable, Person, Any>(requestName: PersonAPI.kRequestPostPersonUpdate,
                                                                                  requestPayload: updatePersonRequestData,
                                                                                  respClass: Person.self,
                                                                                  responseItemType: nil,
                                                                                  responseCompletionHandler: responseCompletionHandler)
        requestProcessor.execute()
    }

    override func findPersonsByStatus(statusList: [Status],
                                      responseCompletionHandler: @escaping ElectrodeBridgeResponseCompletionHandler) {
        let requestProcessor = ElectrodeRequestProcessor<Bridgeable, Array<Any>, Any>(requestName: PersonAPI.kRequestFindPersonsByStatus,
                                                                                      requestPayload: statusList,
                                                                                      respClass: Array<Any>.self,
                                                                                      responseItemType: Person.self,
                                                                                      responseCompletionHandler: responseCompletionHandler)
        requestProcessor.execute()
    }

    override func findPersonByAge(names: [String],
                                  responseCompletionHandler: @escaping ElectrodeBridgeResponseCompletionHandler) {
        let requestProcessor = ElectrodeRequestProcessor<Bridgeable, Array<Any>, Any>(requestName: PersonAPI.kRequestFindPersonsByName,
                                                                                      requestPayload: names,
                                                                                      respClass: Array<Any>.self,
                                                                                      responseItemType: Int.self,
                                                                                      responseCompletionHandler: responseCompletionHandler)
        requestProcessor.execute()
    }
}
